src/host/_stream.cpp

```diff
@@ -228,31 +271,102 @@ void WriteCharsLegacy(SCREEN_INFORMATION& screenInfo, const std::wstring_view& t
+                const auto result = MultiByteToWideChar(cp, MB_USEGLYPHCHARS, &ch, 1, &wch, 1);
+                if (result != 1)
+                {
+                    wch = 0;
+                }
+                if (wch)
+                {
+                    lastCharWrapped = _writeCharsLegacyUnprocessed(screenInfo, { &wch, 1 }, psScrollY);
+                }
                 break;
             }
+            }
 
-            // As a special favor to incompetent apps that attempt to display control chars,
-            // convert to corresponding OEM Glyph Chars
-            const auto cp = ServiceLocator::LocateGlobals().getConsoleInformation().OutputCP;
-            const auto ch = gsl::narrow_cast<char>(*it);
-            wchar_t wch = 0;
-            const auto result = MultiByteToWideChar(cp, MB_USEGLYPHCHARS, &ch, 1, &wch, 1);
-            if (result == 1)
+            if (io)
             {
-                _writeCharsLegacyUnprocessed(screenInfo, { &wch, 1 }, psScrollY);
+                if (wch)
+                {
+                    io->WriteUTF16({ &wch, 1 });
+                }
+                if (lastCharWrapped)
+                {
+                    io->WriteUTF8(" \r");
+                }
+            }
+
+            ++it;
+        } while (it != end && controlCharPredicate(*it));
+    }
+}
+
+void WriteCharsVT(SCREEN_INFORMATION& screenInfo, const std::wstring_view& str)
+{
+    auto& gci = ServiceLocator::LocateGlobals().getConsoleInformation();
+    auto& stateMachine = screenInfo.GetStateMachine();
+
+    stateMachine.ProcessString(str);
+
+    if (const auto io = gci.GetVtIoForBuffer(&screenInfo))
+    {
+        const auto cork = io->Cork();
+        const auto& injections = stateMachine.GetInjections();
+        size_t offset = 0;
+
+        for (const auto& injection : injections)
+        {
+            io->WriteUTF16(til::safe_slice_abs(str, offset, injection.offset));
+            offset = injection.offset;
+
+            switch (injection.type)
+            {
+            case InjectionType::RIS:
+                io->WriteUTF8(
+                    "\033[?1004h" // Focus Event Mode
+                    "\033[?9001h" // Win32 Input Mode
+                );
+                break;
+            case InjectionType::DECSET_FOCUS:
+                io->WriteUTF8(
+                    "\033[?1004h" // Focus Event Mode
+                );
+                break;
             }
         }
+
+        io->WriteUTF16(til::safe_slice_abs(str, offset, std::wstring_view::npos));
+    }
+}
+
+void WriteClearScreen(SCREEN_INFORMATION& screenInfo)
+{
+    std::wstring buffer;
+    buffer.append(L"\033c");
+    if (WI_IsFlagSet(screenInfo.OutputMode, ENABLE_WRAP_AT_EOL_OUTPUT))
+    {
+        buffer.append(L"\x1b[?7h"); // DECAWM: Autowrap Mode
+    }
+    if (WI_IsFlagClear(screenInfo.OutputMode, DISABLE_NEWLINE_AUTO_RETURN))
+    {
+        buffer.append(L"\x1b[20h"); // LNM: Line Feed / New Line Mode
     }

```

Comment:
Injecting the other two isn't a problem because those are specific to ConPTY as it is today (sort of like a multiplexer). It'll always need a VT parser, local buffers, and convert stuff like console attributes to VT.

But in order for us to truly get Windows into a VT world, we somehow need to be able to stop parsing VT sequences, since we shouldn't be the authority on that. Ideally there shouldn't be any changes between what conhost sees and what alacritty or wezterm get and there should not be a VT parser in between.

I wouldn't mind stupidly replacing LF with CRLF but that breaks DCS doesn't it? I'd almost rather enable LNM and replace CRLF with CR in the stdin stream instead. That would net us the issue of not knowing whether a CRLF is the result of pressing Enter or whether it's a genuine CR LF, but I wonder if that's the lesser of the two evils (= vs. parsing VT output in conhost always and forever).

I wonder if there are other solutions...

Reply:
> I wouldn't mind stupidly replacing LF with CRLF but that breaks DCS doesn't it?

Actually I think DCS sequences would probably be fine. I'm fairly sure control characters are supposed to be ignored inside data strings. There might be some non-standard OSC sequences that could be effected, but probably not in a significantly breaking way.

> I'd almost rather enable LNM and replace CRLF with CR in the stdin stream instead.

I considered that as well, but I thought that wouldn't be practical. You'd have to delay on ^M waiting for a potential ^J that may or may not come, which can be flaky on a slow connection.

I don't like either solution really. As you say there might be some other way we can handle this. Need to think about it some more.

Reply:
If can't come up with a better solution I'd be fine with "stupidly replacing LF with CRLF" and saying "If this breaks your VT app, set `DISABLE_NEWLINE_AUTO_RETURN`". The way I see it, and in my understanding, that's no different from setting line disciplines on UNIX, which I suspect can similarly break stuff if not set to a raw mode.